blocks/serialization.py

```diff
@@ -57,41 +58,18 @@ class PersistentParameterID(PersistentNdarrayID):
         `allow_duplicates` is ``False``.
 
     """
-    def __init__(self, zip_file, allow_unnamed=True, allow_duplicates=True):
-        super(PersistentParameterID, self).__init__(zip_file)
-        self.name_counter = defaultdict(int)
-        self.ndarray_names = {}
-        self.allow_unnamed = allow_unnamed
-        self.allow_duplicates = allow_duplicates
-
-    def _resolve_name(self, obj):
-        if id(obj) in self.ndarray_names:
-            name = self.ndarray_names[id(obj)]
-            count = self.name_counter[name]
-            if count:
-                if not self.allow_duplicates:
-                    raise ValueError("multiple shared variables with the name "
-                                     "`{0}` found".format(name))
-                name = '{0}_{1}'.format(name, count + 1)
-            self.name_counter[name] += 1
-            return name
-
     def __call__(self, obj):
         if isinstance(obj, SharedVariable):
-            if obj.name:
-                if obj.name == 'pkl':
-                    raise ValueError(
-                        "can't pickle shared variable with name `pkl`")
-                if hasattr(obj.tag, 'annotations'):
-                    name = BRICK_DELIMITER.join(
-                        [brick.name for brick
-                         in obj.tag.annotations[0].get_unique_path()] +
-                        [obj.name])
-                else:
-                    name = obj.name
-                self.ndarray_names[id(obj.container.storage[0])] = name
-            elif not self.allow_unnamed:
-                raise ValueError("unnamed shared variable, {}".format(obj))
+            super(PersistentParameterID, self).__call__(obj)
+            if hasattr(obj.tag, 'annotations'):
+                name = BRICK_DELIMITER.join(
+                    [brick.name for brick
+                     in get_brick(obj).get_unique_path()] + [obj.name])
+            else:
+                name = obj.name
+            self.ndarray_names[id(obj.container.storage[0])] = name
+        if id(obj) in self.ndarray_names:
+            PersistentCudaNdarrayID.__call__(self, obj)
```

Comment:
That's not what it should do. The only way `id(obj) in self.ndarray_names` can be `True`, is if the `ndarray` or `CudaNdarray` belonged to a shared variable. In that case it will be saved to the ZIP file. For any other array, `id(obj) in self.ndarray_names` will be `False` and the function will return `None`, which tells the pickler to pickle it as usual.